In [28]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re
import collections
from collections import defaultdict
import math

# Loading the data

In [29]:
df_tweets = pd.read_csv("C:\\Users\\USER\\Documents\\Geetha\\Data Science\\Capstone Project 1\\tweets.csv", parse_dates=True)  

In [30]:
df_tweets = df_tweets.set_index(['time','name','username'])

In [31]:
df_tweets.head(20)


description  \
time            name          username                                                        
1/6/2015 21:07  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   

                                              location  followers  \
time            name          username                              
1/6/2015 21:07  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70      NaN        640   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70      NaN        640   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70      NaN        640   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70      NaN        640   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70      NaN        640   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70      NaN        640   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70      NaN        640   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70      NaN        640   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70      NaN        640   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70      NaN        640   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70      NaN        640   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70      NaN        640   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70      NaN        640   

                                               numberstatuses  \
time            name          username                          
1/6/2015 21:07  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:51 

In [32]:
df_tweets.info()


<class 'pandas.core.frame.DataFrame'>
MultiIndex: 17410 entries, (1/6/2015 21:07, GunsandCoffee, GunsandCoffee70) to (5/13/2016 16:10, one who believes, ro34th)
Data columns (total 5 columns):
description       14728 non-null object
location          11432 non-null object
followers         17410 non-null int64
numberstatuses    17410 non-null int64
tweets            17410 non-null object
dtypes: int64(2), object(3)
memory usage: 860.7+ KB


#### Columns 'description' and 'location' have null values

In [33]:
df_tweets['description'] = df_tweets['description'].fillna(' ')

In [34]:
df_tweets['location'] = df_tweets['location'].fillna(' ')

#### NaNs replaced with empty string in the columns 'description' and 'location' 

In [35]:
df_tweets.head(20)

description  \
time            name          username                                                        
1/6/2015 21:07  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70  ENGLISH TRANSLATIONS: http://t.co/QLdJ0ftews   

                                              location  followers  \
time            name          username                              
1/6/2015 21:07  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 21:51  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:04  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:06  GunsandCoffee GunsandCoffee70                 640   
1/6/2015 22:17  GunsandCoffee GunsandCoffee70                 640   
1/10/2015 0:05  GunsandCoffee GunsandCoffee70                 640   
1/10/2015 0:08  GunsandCoffee GunsandCoffee70                 640   
1/12/2015 23:10 GunsandCoffee GunsandCoffee70                 640   
1/22/2015 20:14 GunsandCoffee GunsandCoffee70                 640   
1/22/2015 20:47 GunsandCoffee GunsandCoffee70                 640   
2/9/2015 10:54  GunsandCoffee GunsandCoffee70                 640   
2/12/2015 13:03 GunsandCoffee GunsandCoffee70                 640   
2/12/2015 13:11 GunsandCoffee GunsandCoffee70                 640   
2/16/2015 15:57 GunsandCoffee GunsandCoffee70                 640   
2/20/2015 16:28 GunsandCoffee GunsandCoffee70                 640   
2/20/2015 16:29 GunsandCoffee GunsandCoffee70                 640   

                                               numberstatuses  \
time            name          username                          
1/6/2015 21:07  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:27  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:29  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:37  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:45  GunsandCoffee GunsandCoffee70              49   
1/6/2015 21:51 

In [36]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 17410 entries, (1/6/2015 21:07, GunsandCoffee, GunsandCoffee70) to (5/13/2016 16:10, one who believes, ro34th)
Data columns (total 5 columns):
description       17410 non-null object
location          17410 non-null object
followers         17410 non-null int64
numberstatuses    17410 non-null int64
tweets            17410 non-null object
dtypes: int64(2), object(3)
memory usage: 860.7+ KB


In [37]:
df_tweets=df_tweets.drop_duplicates()

In [38]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 17410 entries, (1/6/2015 21:07, GunsandCoffee, GunsandCoffee70) to (5/13/2016 16:10, one who believes, ro34th)
Data columns (total 5 columns):
description       17410 non-null object
location          17410 non-null object
followers         17410 non-null int64
numberstatuses    17410 non-null int64
tweets            17410 non-null object
dtypes: int64(2), object(3)
memory usage: 860.7+ KB


In [39]:
df_tweets.columns


Index(['description', 'location', 'followers', 'numberstatuses', 'tweets'], dtype='object')

In [40]:
df_tweets.index.get_level_values(1).value_counts(normalize=True)



Rami                    0.084721
War BreakingNews        0.068409
Conflict Reporter       0.062895
Salahuddin Ayubi        0.060655
Ibni Haneefah           0.040724
                          ...   
Khilafah Superstate     0.000402
Ibn Al-Shishani         0.000287
Abdul05                 0.000287
Newerajihadi61          0.000230
fahad slyman alsomal    0.000230
Name: name, Length: 112, dtype: float64

#### We are analyzing the tweets of 112 different users who have generated 17410 tweets

In [41]:
df_tweets['description'] = df_tweets['description'].str.lower()
df_tweets['tweets'] = df_tweets['tweets'].str.lower()
#df_tweets.head(20)


In [42]:

#df_tweets['tweets'] =  df_tweets['tweets'] .apply(lambda elem: re.sub(r"(!|@|#|\n|')", "", elem))

tweet_tokens = [re.sub (r'http\S+', 'url', x) for x in df_tweets['tweets']]
tweet_tokens = [re.sub (r"(!|#|\n|'|''|:|,|[0-9]|""|@\S+|/|-)", "", x) for x in tweet_tokens ]

tweet_tokens

['english translation a message to the truthful in syria  sheikh abu muhammed al maqdisi url url',
 'english translation sheikh fatih al jawlani for the people of integrity sacrifice is  easy url url',
 'english translation first audio meeting with sheikh fatih al jawlani (ha) url url',
 'english translation sheikh nasir al wuhayshi (ha) leader of aqap the promise of victory url url',
 'english translation aqap response to sheikh baghdadis statement although the disbelievers dislike it. url',
 'the second clip in a dawah series by a soldier of jn video link url url',
 'english transcript  oh murabit  url url',
 'english translation a collection of the words of the ulama regarding dawlah url url',
 'aslm please share our new account after the previous one was suspended.  ',
 'english translation aqap statement regarding the blessed raid in franceurl url',
 '   ',
 'new link after previous one taken downaqapthe faces have been brightened regarding the blessed attack in france url',
 'eng

#### Removing some of the special characters from 'tweets'column using regular expression operations

In [43]:
#tweet_tokens3=['english translation a', 'english translation']

tweet_tokens_split =[x.split(" ") for x in tweet_tokens]
for single_list in tweet_tokens_split:
   for single_word in single_list:
       if len(single_word) <= 1:
          single_list.remove(single_word)
#tweet_tokens3 =[' '.join(words) for words in tweet_tokens3]


tweet_tokens_split

[['english',
  'translation',
  'message',
  'to',
  'the',
  'truthful',
  'in',
  'syria',
  'sheikh',
  'abu',
  'muhammed',
  'al',
  'maqdisi',
  'url',
  'url'],
 ['english',
  'translation',
  'sheikh',
  'fatih',
  'al',
  'jawlani',
  'for',
  'the',
  'people',
  'of',
  'integrity',
  'sacrifice',
  'is',
  'easy',
  'url',
  'url'],
 ['english',
  'translation',
  'first',
  'audio',
  'meeting',
  'with',
  'sheikh',
  'fatih',
  'al',
  'jawlani',
  '(ha)',
  'url',
  'url'],
 ['english',
  'translation',
  'sheikh',
  'nasir',
  'al',
  'wuhayshi',
  '(ha)',
  'leader',
  'of',
  'aqap',
  'the',
  'promise',
  'of',
  'victory',
  'url',
  'url'],
 ['english',
  'translation',
  'aqap',
  'response',
  'to',
  'sheikh',
  'baghdadis',
  'statement',
  'although',
  'the',
  'disbelievers',
  'dislike',
  'it.',
  'url'],
 ['the',
  'second',
  'clip',
  'in',
  'dawah',
  'series',
  'by',
  'soldier',
  'of',
  'jn',
  'video',
  'link',
  'url',
  'url'],
 ['english',

#### Removing single character words

In [44]:

frequency_list = defaultdict(int)
for word_list in tweet_tokens_split:
   for word in word_list:
       #if single_word in single_list:
          frequency_list[word]+=1
          #word_list.remove(word)  
sorted(frequency_list.items(), key=lambda x: x[1],reverse=True)



[('url', 8622),
 ('the', 7382),
 ('in', 6604),
 ('of', 5924),
 ('rt', 5851),
 ('to', 4199),
 ('and', 3404),
 ('is', 3353),
 ('isis', 2132),
 ('', 2057),
 ('on', 1735),
 ('by', 1672),
 ('for', 1514),
 ('with', 1362),
 ('&amp;', 1347),
 ('from', 1288),
 ('syria', 1271),
 ('you', 1242),
 ('are', 1230),
 ('killed', 1224),
 ('they', 1125),
 ('that', 981),
 ('this', 951),
 ('army', 864),
 ('it', 831),
 ('us', 790),
 ('was', 762),
 ('not', 753),
 ('have', 747),
 ('their', 699),
 ('who', 673),
 ('will', 663),
 ('after', 652),
 ('an', 648),
 ('islamic', 636),
 ('state', 611),
 ('be', 610),
 ('as', 607),
 ('near', 566),
 ('he', 565),
 ('against', 562),
 ('at', 553),
 ('assad', 539),
 ('iraq', 530),
 ('allah', 518),
 ('its', 511),
 ('but', 492),
 ('soldiers', 492),
 ('attack', 478),
 ('has', 470),
 ('them', 452),
 ('al', 451),
 ('forces', 451),
 ('if', 450),
 ('aleppo', 449),
 ('his', 447),
 ('were', 440),
 ('one', 429),
 ('today', 417),
 ('syrian', 397),
 ('breaking', 391),
 ('people', 388),
 ('

#### Finding the most frequent words in 'tweets' column inorder to create a list of stop words

In [45]:

idf = defaultdict(int)
for word_list in tweet_tokens_split:
   for word in word_list:
       #if single_word in single_list:
          idf[word]+=1
          word_list.remove(word) 

count_tweets= df_tweets.tweets.count()
for key, value in idf.items():
    idf[key]=math.log(count_tweets/(value + 1))
sorted(idf.items(), key=lambda x: x[1],reverse=False)


[('rt', 1.0921427387219254),
 ('url', 1.3940208598015336),
 ('the', 1.5672612930410508),
 ('in', 1.6646385858256272),
 ('of', 1.787174933977642),
 ('to', 2.1094093879360827),
 ('and', 2.3317246838636545),
 ('is', 2.3430242391175877),
 ('', 2.5809293176997823),
 ('isis', 2.7708670575390455),
 ('on', 2.9513554332513388),
 ('by', 3.0101959332742725),
 ('for', 3.108073508583843),
 ('&amp;', 3.2080216766041922),
 ('with', 3.2353811945000093),
 ('from', 3.2755951014369176),
 ('you', 3.303331856408517),
 ('killed', 3.344805104615092),
 ('are', 3.3595715747313935),
 ('they', 3.3779207133995897),
 ('syria', 3.379605633764509),
 ('this', 3.5663213162699265),
 ('that', 3.6078210471766794),
 ('it', 3.7245453214848205),
 ('army', 3.73893405893692),
 ('not', 3.746206818266),
 ('us', 3.7808637520750445),
 ('was', 3.8540033887217073),
 ('have', 3.8649026791797434),
 ('who', 3.892682243286819),
 ('after', 3.898331975828938),
 ('their', 3.944717102409873),
 ('islamic', 3.971786424378091),
 ('will', 4.01

#### Finding the low IDF(Inverse Document Frequency ) terms to add them to the list of stop words

In [46]:
'''
tweet_tokens_joined = [item for sublist in tweet_tokens_split for item in sublist]
count_stopwords = collections.Counter(tweet_tokens_joined)
sorted(count_stopwords.items(), key=lambda x: x[1],reverse=True)
'''



'\ntweet_tokens_joined = [item for sublist in tweet_tokens_split for item in sublist]\ncount_stopwords = collections.Counter(tweet_tokens_joined)\nsorted(count_stopwords.items(), key=lambda x: x[1],reverse=True)\n'

In [47]:
stop_words=['url','the', 'in', 'of', 'to', 'and', 'is', 'on', 'by', 'for', 'with', '&amp;', 'from', 'are', 'killed', 'you', 'they', 'that', 'this', 'it', 'us', 'was', 'have', 'their', 'will', 'an', 'who', 'be', 'as', 'after', 'at', 'he', 'al', 'but', 'its', 'has', 'his', 'one', 'were', 'if', 'all', 'today', 'them', 'we', 'city', 'people', 'about', 'your', 'now', 'when', 'ypg', 'what', 'more', 'or', 'new', 'over', 'like', 'just', 'so', 'can', 'de', 'west', 'north', 'saa','how', 'only', 'do', 'claims', 'reports', 'those', 'captured', 'fsa', 'than', 'my', 'back', 'up', 'la', 'between', 'group', 'our', 'there', 'out', 'two', 'area', 'while', 'via', 'which', 'know', 'because', 'been','see', 'le', 'coalition', 'me', 'village', 'south', 'northern', 'news', 'many', 'him', 'east', 'time', 'being', 'some', 'still', 'these', 'think', 'support', 'then', 'may', 'a', 'why', 'make', 'air','take', 'also', 'want', 'huge', 'says', 'even', 'un', 'under', 'during', 'other', 'very', 'homs','say', 'les', 'targeted', 'help', 'into','days', 'reportedly', 'eastern', 'any','said','please', 'des', 'where', 'would', 'first', 'yesterday', 'get', 'ied', 'heavy', 'report', 'day', 'security', 'same', 'twitter','im', 'ahrar', 'least', 'htt…', 'years', 'cest', 'deir','je', 'vso','rt', 'url', 'the', 'in', 'of', 'to', 'and', 'is', '', 'on', 'by', 'for', '&amp;', 'with', 'from', 'you', 'are', 'they', 'this', 'that', 'it', 'was', 'have', 'who', 'after', 'their', 'will', 'as', 'an', 'be', 'he', 'near', 'its', 'at', 'but', 'them', 'has', 
'if', 'were', 'his', 'al', 'english', 'translation', 'today', 'one', 'when', 'people', 'or', 'city', 'what', 'we', 'now', 'so', 'all', 'more', 'only','about', 'your', 'ypg', 'de', 'new', 'reports', 'just', 'over', 'saa', 'like', 'my', 'do', 'i', 'can', 'area', 'there','up', 'between', 'how', 'our', 'than', 'claims', 'group', 'out', 'may', 'la', 'huge', 'those', 'me', 'some', 'le', 'time', 'see', 'been', 'know', 'make', 'many', 'him', 'still', 'which', 'why', 'clashes', 'says', 'two', 'south', 'because', 'via', 'these', 'support', 'being', 'then', 'un', 'other', 'said', 'control', 'back', 'under', 'very', 'where', 'while', 'any', 'say', 'even', 'air', 'lol', 'during', 'eastern', 'get', 'days', 'want', 'please', 'also', 'most','first', 'im', 'into', 'another', 'les', 'would', 'positions', 'reportedly']

stop_words = list(set(stop_words))



In [48]:
for word_list in tweet_tokens_split:
      for word in word_list:
       if word in stop_words:          
          word_list.remove(word)
   
tweet_tokens_split

[['to', 'truthful', 'syria', 'abu', 'url'],
 ['fatih', 'jawlani', 'of', 'sacrifice', 'easy'],
 ['audio', 'fatih', 'jawlani'],
 ['nasir', 'wuhayshi', 'leader', 'aqap', 'promise', 'victory'],
 ['response', 'sheikh', 'statement', 'dislike'],
 ['second', 'series', 'soldier', 'jn', 'link'],
 ['transcript', 'murabit'],
 ['words', 'the', 'regarding'],
 ['our', 'account', 'one', 'suspended.'],
 ['statement', 'raid', 'franceurl'],
 [],
 ['link', 'previous', 'taken', 'faces', 'regarding', 'blessed', 'url'],
 ['abu', '(ha)', 's.a.w', 'url'],
 ['abu', '(ha)', 'saw', 'url'],
 ['not'],
 ['anybody', 'jn', 'it', 'nobody', ''],
 ['we', 'translate', 'let', 'knw', 'anybody', 'available'],
 ['half', 'through', 'inshallah'],
 ['we', 'completed', 'the', 'news'],
 ['and', 'syrian', 'that'],
 ['should', 'it', 'not?'],
 ['links', 'jn', 'eng', 'of', 'watchurlurl'],
 ['jn', 'subtitles', 'path', 'salvation'],
 ['eng', 'save', 'watch', 'download'],
 ['released', 'jn', 'of', 'islamic', 'regarding'],
 ['anybody', 'a

#### Removing stopwords from the list of words in 'tweets' column

In [49]:
list_words= ['turkey','islamicstate','syria','aleppo','warreporter1','russia', 'assad', 'iraq','url']
dict_repetition=defaultdict(set)
for item in list_words:
   for word_list in tweet_tokens_split:
      for word in word_list:
          if item in word:          
              dict_repetition[item].add(word)

for keys in dict_repetition: 
    dict_repetition[keys] = tuple(dict_repetition[keys]) 

dict_inverse = {v: k for k, v in dict_repetition.items()} 
dict_inverse


{('proturkey',
  'sauditurkey',
  'turkeys',
  'breakingturkey',
  'youturkey',
  'turkey;',
  'usturkeysponsored',
  'turkey..',
  'turkey',
  'docxurlturkey',
  'turkey.',
  'ankara.turkeys',
  'days.turkeyurl',
  'turkey’s',
  'turkeybacked',
  'turkey.pkk',
  'ksaturkeyqatars',
  'turkeyis…'): 'turkey',
 ('ramadi..islamicstate',
  'aamaqislamicstate',
  'footageislamicstate',
  'islamicstatewilayatninawaglimpse',
  'islamicstate...',
  'kirkukislamicstate',
  'iraqislamicstate',
  'mosulislamicstate',
  'islamicstatecontrolled',
  'islamicstateyaldayarmoukdamascus',
  'saaislamicstate',
  'abundance"||islamicstate',
  'islamicstatelove',
  'islamicstate.may',
  'agencyislamicstate',
  'damascusislamicstate...',
  'libyaislamicstate',
  'islamicstatewilayatdimashq',
  'islamicstate..',
  'islamicstatelost',
  'amaqislamicstate',
  'islamicstatewilayatkhurasan',
  'syriaislamicstate',
  'islamicstate',
  'islamicstatecouncil',
  'islamicstateurl',
  'belgiumupdateislamicstate',
  '"|

#### Creating a dictionary for words with same root

In [50]:
for word_list in tweet_tokens_split:
    for i,word in enumerate(word_list):
        for key,value in dict_inverse.items():
            if word in key:
               word_list[i]=value


tweet_tokens_split

[['to', 'truthful', 'syria', 'abu', 'url'],
 ['fatih', 'jawlani', 'of', 'sacrifice', 'easy'],
 ['audio', 'fatih', 'jawlani'],
 ['nasir', 'wuhayshi', 'leader', 'aqap', 'promise', 'victory'],
 ['response', 'sheikh', 'statement', 'dislike'],
 ['second', 'series', 'soldier', 'jn', 'link'],
 ['transcript', 'murabit'],
 ['words', 'the', 'regarding'],
 ['our', 'account', 'one', 'suspended.'],
 ['statement', 'raid', 'url'],
 [],
 ['link', 'previous', 'taken', 'faces', 'regarding', 'blessed', 'url'],
 ['abu', '(ha)', 's.a.w', 'url'],
 ['abu', '(ha)', 'saw', 'url'],
 ['not'],
 ['anybody', 'jn', 'it', 'nobody', ''],
 ['we', 'translate', 'let', 'knw', 'anybody', 'available'],
 ['half', 'through', 'inshallah'],
 ['we', 'completed', 'the', 'news'],
 ['and', 'syria', 'that'],
 ['should', 'it', 'not?'],
 ['links', 'jn', 'eng', 'of', 'url'],
 ['jn', 'subtitles', 'path', 'salvation'],
 ['eng', 'save', 'watch', 'download'],
 ['released', 'jn', 'of', 'islamic', 'regarding'],
 ['anybody', 'audio', 'sheikh'

#### Replacing the words in the 'tweets' column with their root

In [51]:
tweet_tokens_joined =[' '.join(words) for words in tweet_tokens_split]
tweet_tokens_joined

['to truthful syria abu url',
 'fatih jawlani of sacrifice easy',
 'audio fatih jawlani',
 'nasir wuhayshi leader aqap promise victory',
 'response sheikh statement dislike',
 'second series soldier jn link',
 'transcript murabit',
 'words the regarding',
 'our account one suspended.',
 'statement raid url',
 '',
 'link previous taken faces regarding blessed url',
 'abu (ha) s.a.w url',
 'abu (ha) saw url',
 'not',
 'anybody jn it nobody ',
 'we translate let knw anybody available',
 'half through inshallah',
 'we completed the news',
 'and syria that',
 'should it not?',
 'links jn eng of url',
 'jn subtitles path salvation',
 'eng save watch download',
 'released jn of islamic regarding',
 'anybody audio sheikh it no else',
 'will',
 'rules imarah conditions url definitions translation…',
 'with tartoosi',
 'الإثنين هـ صوتيّة مكتوبة',
 'confirmed state aljusiya post jurud to countryside',
 'alaykum. back. supports spread barakallahu',
 '',
 'generation takbirrr dawlah',
 'akhi',
 'uk

#### Tf-idf

In [52]:
#Import TfidfVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer 
#max_features=1000, stop_words=stop_words_list
# Instantiate TfidfVectorizer
tv = TfidfVectorizer( max_features=1000,) 

# Fit the vectroizer and transform the data
tv_transformed = tv.fit_transform(tweet_tokens_joined )

# Create a DataFrame with these features
tv_df = pd.DataFrame(tv_transformed.toarray(), 
                     columns=tv.get_feature_names()).add_prefix('TFIDF_')
#tv.vocabulary_
tv.get_feature_names()
#tv_df

['aamaq',
 'able',
 'about',
 'abu',
 'accept',
 'according',
 'account',
 'accounts',
 'advance',
 'afghan',
 'afghanistan',
 'africa',
 'after',
 'aftermath',
 'again',
 'against',
 'agency',
 'ago',
 'agree',
 'ahmad',
 'ahrar',
 'aid',
 'air',
 'airbase',
 'aircraft',
 'airport',
 'airstrike',
 'airstrikes',
 'akbar',
 'akhi',
 'al',
 'aleppo',
 'alert',
 'ali',
 'alislam',
 'alive',
 'all',
 'allah',
 'allahu',
 'allegiance',
 'allied',
 'allies',
 'alloush',
 'allowed',
 'ally',
 'allāh',
 'alnusra',
 'alone',
 'alqaeda',
 'already',
 'also',
 'always',
 'am',
 'amaq',
 'amaqagency',
 'amaqnewsagency',
 'ameen',
 'america',
 'american',
 'among',
 'amp',
 'an',
 'anbar',
 'and',
 'ankara',
 'answer',
 'anwar',
 'anyone',
 'anything',
 'apostate',
 'apostates',
 'aq',
 'aqap',
 'arab',
 'arabia',
 'arabic',
 'arabs',
 'are',
 'area',
 'areas',
 'armed',
 'army',
 'around',
 'arrested',
 'article',
 'artillery',
 'as',
 'ask',
 'assad',
 'assault',
 'at',
 'attack',
 'attacked',
 '

In [53]:
# Isolate the row to be examined
sample_row = tv_df.iloc[0]

# Print the top 5 words of the sorted output
print(sample_row.sort_values(ascending=False).head(10))


TFIDF_abu        0.641061
TFIDF_to         0.500469
TFIDF_syria      0.435158
TFIDF_url        0.386276
TFIDF_full       0.000000
TFIDF_freedom    0.000000
TFIDF_french     0.000000
TFIDF_from       0.000000
TFIDF_front      0.000000
TFIDF_fsa        0.000000
Name: 0, dtype: float64


In [54]:
tfidf = TfidfVectorizer(min_df=3)
tfidf.fit(tweet_tokens_joined)
feature_names = tfidf.get_feature_names()
feature_names

['aafia',
 'aamaq',
 'aamaqis',
 'ab',
 'abadi',
 'abandon',
 'abandoned',
 'abbas',
 'abdelhaleem',
 'abdeslam',
 'abdul',
 'abdullah',
 'abi',
 'ability',
 'able',
 'abo',
 'about',
 'above',
 'abraham',
 'abrams',
 'abrini',
 'abroad',
 'absolutely',
 'abt',
 'abu',
 'abu_ghraib',
 'abyad',
 'acc',
 'accept',
 'accepted',
 'accepts',
 'according',
 'account',
 'accountable',
 'accounts',
 'accurate',
 'accusation',
 'accuse',
 'accused',
 'achieved',
 'acount',
 'across',
 'act',
 'action',
 'actions',
 'active',
 'activist',
 'activists',
 'activities',
 'activity',
 'acts',
 'actual',
 'actually',
 'ad',
 'add',
 'added',
 'additional',
 'address',
 'aden',
 'administration',
 'admits',
 'admitted',
 'adnani',
 'adopted',
 'adultery',
 'advance',
 'advances',
 'advancing',
 'advice',
 'adviser',
 'advisor',
 'aerial',
 'afar',
 'affairs',
 'afghan',
 'afghanistan',
 'afghans',
 'afp',
 'africa',
 'african',
 'after',
 'aftermath',
 'again',
 'against',
 'age',
 'agencies',
 'agenc

In [55]:
def get_ifidf_for_words(text):
    tfidf_matrix= tfidf.transform([text]).todense()
    feature_index = tfidf_matrix[0,:].nonzero()[1]
    tfidf_scores = zip([feature_names[i] for i in feature_index], [tfidf_matrix[0, x] for x in feature_index])
    return dict(tfidf_scores)

In [56]:
get_ifidf_for_words(tweet_tokens_joined[0])

{'abu': 0.439549339839888,
 'syria': 0.29837036374828957,
 'to': 0.3431509292291457,
 'truthful': 0.7279226059345042,
 'url': 0.2648539662186054}